### ISSUE TITLE
Test Failure

### ISSUE THREAD

**Issue by toumorokoshi** [toumorokoshi/tome#6]
👋 (this is spawned issue from #2, which has become a bigger discussion)

Long time `sub` fan and found your project thanks to a friend. Thanks for setting up a new incarnation! I've setup a `sub` based cli at the last couple places I've worked and it's been great. Interested to play with tome and see if it's my new goto.

When trying to run the current HEAD of tome, I get a test failure:

```
bash-5.1$ cargo test
    Finished test [unoptimized + debuginfo] target(s) in 0.01s
     Running target/debug/deps/tome-2fa1d8bb501aff4b

running 16 tests
test commands::tests::test_should_not_source ... ok
test commands::tests::test_usage ... ok
test commands::tests::test_should_source ... ok
test commands::tests::test_help ... ok
test tests::test_dangerous_characters_quoted ... ok
test tests::test_script_directory_argument ... ok
test tests::test_directory_completion ... ok
test tests::test_script_in_directory ... ok
test tests::test_script_in_directory_not_found ... ok
test tests::test_root_directory_completion ... ok
test tests::test_simple_script ... ok
test tests::test_source ... ok
test tests::test_use_arg ... ok
test tests::test_help_page ... ok
test tests::test_simple_script_completion ... ok
test tests::test_source_completion ... FAILED

failures:

---- tests::test_source_completion stdout ----
thread 'tests::test_source_completion' panicked at 'assertion failed: `(left == right)`
  left: `Ok("")`,
 right: `Ok("foo baz\n")`', src/tests.rs:43:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::test_source_completion

test result: FAILED. 15 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.17s
```

I investigated this ~by setting up a script to recompile tome and prepend the compiled ./target/debug/tome` to my path (otherwise system tome would be used) and~ (edited due to my misunderstanding test harness) re-run the tests across the various git tags. In my testing this one test has failed from v0.1 onwards to v0.7.1.

A couple questions:
* Are you seeing different test results than I am?
* Is the desired behavior still that `source` style scripts allow for completion?
* If I'm hacking on this, would you appreciate things like Github Actions for CI and patches for that test failure?

**Comment by toumorokoshi** [comment]
I just tried this on my local machine, it passes (Arch Linux, rust stable):

```
running 16 tests
test commands::tests::test_should_not_source ... ok
test commands::tests::test_help ... ok
test commands::tests::test_should_source ... ok
test commands::tests::test_usage ... ok
test tests::test_dangerous_characters_quoted ... ok
test tests::test_directory_completion ... ok
test tests::test_script_directory_argument ... ok
test tests::test_script_in_directory ... ok
test tests::test_simple_script ... ok
test tests::test_script_in_directory_not_found ... ok
test tests::test_root_directory_completion ... ok
test tests::test_use_arg ... ok
test tests::test_source ... ok
test tests::test_help_page ... ok
test tests::test_source_completion ... ok
test tests::test_simple_script_completion ... ok
```

Thanks for the output on yours. @zph could this be a difference in the default shells? what shell are you using to run your tests?

**Comment by zph** [comment]
Peculiar... could be shell specific.

Debugging...

On Big Sur OSX, using Alacritty Terminal, with fish shell v3.1.2.

```
❯ rustc --version                                                                                               13:51:37
rustc 1.50.0 (cb75ad5db 2021-02-10)
```


Using `upstream/master` aka f4e2825 and run the following commands:

```
cargo clean
cargo test
```

I receive the following failure.
<img width="727" alt="image" src="https://user-images.githubusercontent.com/1026584/111885105-e3785500-8982-11eb-982e-7f474b3f2211.png">

When I run `cargo clean && cargo test` in bash 5.1, I receive the same failure. Then same failure in zsh.

Then I reverted the test deletion and removed the `ignore` tag and pushed it so we can see what Github actions says for `cargo test`. It fails there :( (https://github.com/zph/tome/runs/2156928152).

I'm a bit stumped. Can you try `cargo clean && cargo test` in case something's stale?

**Comment by toumorokoshi** [comment]
Just outputting data, no insights yet:

```
tsutsumi@tsutsumi-laptop ~/w/tome (master)> rustc --version
rustc 1.50.0 (cb75ad5db 2021-02-10)
```

I'm running Arch Linux, zsh:

```
tsutsumi:/home/tsutsumi/workspace/tome (0) (master)
$ echo $0
/bin/zsh
tsutsumi:/home/tsutsumi/workspace/tome (0) (master)
$ cargo clean && cargo test
   Compiling libc v0.2.48
   Compiling cfg-if v0.1.6
   Compiling either v1.5.0
   Compiling cassowary v0.3.0
   Compiling bitflags v1.0.4
   Compiling unicode-width v0.1.5
   Compiling unicode-segmentation v1.2.1
   Compiling log v0.4.6
   Compiling itertools v0.7.11
   Compiling termion v1.5.1
   Compiling tui v0.3.0
   Compiling tome v0.1.0 (/home/tsutsumi/workspace/tome)
    Finished test [unoptimized + debuginfo] target(s) in 3.94s
     Running target/debug/deps/tome-6a43bb37b00093a7

running 16 tests
test commands::tests::test_should_source ... ok
test commands::tests::test_help ... ok
test commands::tests::test_should_not_source ... ok
test commands::tests::test_usage ... ok
test tests::test_dangerous_characters_quoted ... ok
test tests::test_script_directory_argument ... ok
test tests::test_directory_completion ... ok
test tests::test_root_directory_completion ... ok
test tests::test_script_in_directory ... ok
test tests::test_script_in_directory_not_found ... ok
test tests::test_simple_script ... ok
test tests::test_source ... ok
test tests::test_use_arg ... ok
test tests::test_help_page ... ok
test tests::test_source_completion ... ok
test tests::test_simple_script_completion ... ok

test result: ok. 16 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.06s

tsutsumi:/home/tsutsumi/workspace/tome (0) (master)
```

trying with bash...

```
[tsutsumi@tsutsumi-laptop tome]$ cargo clean && cargo test
   Compiling libc v0.2.48
   Compiling either v1.5.0
   Compiling cfg-if v0.1.6
   Compiling bitflags v1.0.4
   Compiling unicode-segmentation v1.2.1
   Compiling cassowary v0.3.0
   Compiling unicode-width v0.1.5
   Compiling log v0.4.6
   Compiling itertools v0.7.11
   Compiling termion v1.5.1
   Compiling tui v0.3.0
   Compiling tome v0.1.0 (/home/tsutsumi/workspace/tome)
    Finished test [unoptimized + debuginfo] target(s) in 3.94s
     Running target/debug/deps/tome-6a43bb37b00093a7

running 16 tests
test commands::tests::test_help ... ok
test commands::tests::test_should_source ... ok
test commands::tests::test_should_not_source ... ok
test commands::tests::test_usage ... ok
test tests::test_dangerous_characters_quoted ... ok
test tests::test_directory_completion ... ok
test tests::test_script_directory_argument ... ok
test tests::test_script_in_directory ... ok
test tests::test_root_directory_completion ... ok
test tests::test_script_in_directory_not_found ... ok
test tests::test_simple_script ... ok
test tests::test_source ... ok
test tests::test_use_arg ... ok
test tests::test_help_page ... ok
test tests::test_source_completion ... ok
test tests::test_simple_script_completion ... ok

test result: ok. 16 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.06s
```

fish:

```
tsutsumi@tsutsumi-laptop ~/w/tome (master)> cargo clean && cargo test
   Compiling libc v0.2.48
   Compiling cfg-if v0.1.6
   Compiling either v1.5.0
   Compiling bitflags v1.0.4
   Compiling unicode-width v0.1.5
   Compiling unicode-segmentation v1.2.1
   Compiling cassowary v0.3.0
   Compiling log v0.4.6
   Compiling itertools v0.7.11
   Compiling termion v1.5.1
   Compiling tui v0.3.0
   Compiling tome v0.1.0 (/home/tsutsumi/workspace/tome)
    Finished test [unoptimized + debuginfo] target(s) in 3.90s
     Running target/debug/deps/tome-6a43bb37b00093a7

running 16 tests
test commands::tests::test_help ... ok
test commands::tests::test_should_not_source ... ok
test commands::tests::test_should_source ... ok
test commands::tests::test_usage ... ok
test tests::test_dangerous_characters_quoted ... ok
test tests::test_directory_completion ... ok
test tests::test_script_directory_argument ... ok
test tests::test_root_directory_completion ... ok
test tests::test_script_in_directory ... ok
test tests::test_simple_script ... ok
test tests::test_source ... ok
test tests::test_use_arg ... ok
test tests::test_help_page ... ok
test tests::test_script_in_directory_not_found ... ok
test tests::test_source_completion ... ok
test tests::test_simple_script_completion ... ok

test result: ok. 16 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.07s

tsutsumi@tsutsumi-laptop ~/w/tome (master)>
```
SHA is same as yours:

```
commit f4e28252cf54d07b003d6dc80586be5bd168b6bc (HEAD -> master, tag: v0.7.2, origin/master, origin/HEAD)
Author: Yusuke Tsutsumi <yusuke@tsutsumi.io>
Date:   Sun Feb 2 20:55:45 2020 -0800
```

So maybe OSX is the one delta? Unfortunately I don't have an OSX machine on hand. But It was time to buy one anyway :)

**Comment by zph** [comment]
Theory... modern versions of OSX don't recent versions of bash in their default setup. Wonder if that is part of our issue 🤔 but don't yet have a theory for `why`.

ie: I have installed a new bash via homebrew:

```
❯ which -a bash                                                                                                 13:55:37
/usr/local/bin/bash # version 5.1.4(1)
/bin/bash                # version 3.2.57(1)
```

I'll see if I can this into a coherent & well reasoned theory.

**Comment by zph** [comment]
Ok, found it.

I added a print statement in `get_execution_body`, here:

```
let cmd = command.args(args).stdout(Stdio::piped())
println!("Cmd: {:#?}", cmd);
let command_output = cmd.output();
```

**When I run this from fish:**

```
   Compiling tome v0.1.0 (/Users/zph/src/tome)
    Finished test [unoptimized + debuginfo] target(s) in 0.97s
     Running target/debug/deps/tome-2fa1d8bb501aff4b

running 16 tests
test commands::tests::test_help ... ok
test commands::tests::test_usage ... ok
test commands::tests::test_should_source ... ok
test commands::tests::test_should_not_source ... ok
test tests::test_script_directory_argument ... ok
test tests::test_script_in_directory_not_found ... ok
test tests::test_directory_completion ... ok
test tests::test_dangerous_characters_quoted ... ok
test tests::test_use_arg ... ok
test tests::test_root_directory_completion ... ok
test tests::test_simple_script ... ok
test tests::test_script_in_directory ... ok
test tests::test_source ... ok
test tests::test_help_page ... ok
test tests::test_simple_script_completion ... ok
test tests::test_source_completion ... FAILED

failures:

---- tests::test_source_completion stdout ----
Cmd: "/usr/local/bin/fish" "./example/source_example" "--complete"
thread 'tests::test_source_completion' panicked at 'assertion failed: `(left == right)`
  left: `Ok("")`,
 right: `Ok("foo baz\n")`', src/tests.rs:43:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::test_source_completion

test result: FAILED. 15 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.23s

error: test failed, to rerun pass '--bin tome'
```

Specifically the log output: `Cmd: "/usr/local/bin/fish" "./example/source_example" "--complete"`

**When I start a terminal in fish, then switch into bash:**
```
bash-5.1$ cargo test
    Finished test [unoptimized + debuginfo] target(s) in 0.01s
     Running target/debug/deps/tome-2fa1d8bb501aff4b

running 16 tests
test commands::tests::test_usage ... ok
test commands::tests::test_help ... ok
test commands::tests::test_should_source ... ok
test commands::tests::test_should_not_source ... ok
test tests::test_dangerous_characters_quoted ... ok
test tests::test_script_directory_argument ... ok
test tests::test_script_in_directory_not_found ... ok
test tests::test_script_in_directory ... ok
test tests::test_directory_completion ... ok
test tests::test_source ... ok
test tests::test_simple_script ... ok
test tests::test_use_arg ... ok
test tests::test_root_directory_completion ... ok
test tests::test_help_page ... ok
test tests::test_simple_script_completion ... ok
test tests::test_source_completion ... FAILED

failures:

---- tests::test_source_completion stdout ----
Cmd: "/usr/local/bin/fish" "./example/source_example" "--complete"
thread 'tests::test_source_completion' panicked at 'assertion failed: `(left == right)`
  left: `Ok("")`,
 right: `Ok("foo baz\n")`', src/tests.rs:43:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::test_source_completion

test result: FAILED. 15 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.15s

error: test failed, to rerun pass '--bin tome'
```

Note that SHELL is still reporting the same: `Cmd: "/usr/local/bin/fish" "./example/source_example" "--complete"`.

I can force my tests to pass by doing:

```
bash-5.1$ SHELL=/bin/bash cargo test
    Finished test [unoptimized + debuginfo] target(s) in 0.06s
     Running target/debug/deps/tome-2fa1d8bb501aff4b

running 16 tests
test commands::tests::test_should_not_source ... ok
test commands::tests::test_should_source ... ok
test commands::tests::test_usage ... ok
test commands::tests::test_help ... ok
test tests::test_script_directory_argument ... ok
test tests::test_script_in_directory_not_found ... ok
test tests::test_directory_completion ... ok
test tests::test_root_directory_completion ... ok
test tests::test_dangerous_characters_quoted ... ok
test tests::test_use_arg ... ok
test tests::test_simple_script ... ok
test tests::test_script_in_directory ... ok
test tests::test_source ... ok
test tests::test_help_page ... ok
test tests::test_source_completion ... ok
test tests::test_simple_script_completion ... ok

test result: ok. 16 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.05s
```

When run from fish, I can force `success` by doing the equivalent:

```
env SHELL=/bin/bash cargo test
```

This addresses why it's failing in my terminal but succeeding in yours. And I suspect the SHELL is set when instantiating the terminal but isn't being updated when I switch shells mid terminal session 🤔.

**Comment by zph** [comment]
I doublechecked the CI test failure... it's for same test but different error msg 😆.

Moral of the story, this test can't be run from fish shell and it's failing on CI for unrelated reasons in same test. I think the actual functionality is working per spec of what you're wanting. But the test isn't reliable across different platforms, ie fish shell, or bash when terminal starts as fish, and CI also has a different error in same test.

This brings up an interface and philosophical question:
- How do we test execute a `source` style script, when we don't know what shell should be used for sourcing? Because bash/zsh and fish are incompatible at each other's scripts.
- We could force override the SHELL variable in test circumstances to ensure reliability 👍 

I'll try a patch like that and then see if I can figure out what's up with Github Actions execution.

**Comment by zph** [comment]
https://github.com/zph/tome/runs/2157772659?check_suite_focus=true

Fixed both test consistency across environments and in CI.

Issue can be closed and fix will merge in #4.

**Comment by zph** [comment]
Fixed in #4 and merged in.

This ticket can be closed.
